application/data/standardisers/ethnicity_classification_finder.py

```diff
@@ -0,0 +1,287 @@
+            "data": self.__get_mapped_raw_data(raw_ethnicities, ethnicity_standardiser),
+        }
+
+    def __get_classification_as_dictionary(self):
+        standards = list(self.standard_value_to_display_value_map.keys())
+        return {
+            "code": self.code,
+            "name": self.name,
+            "map": {
+                standard: self.__get_data_item_for_standard_ethnicity(standard).to_dict() for standard in standards
+            },
+        }
+
+    def __get_mapped_raw_data(self, raw_ethnicities, ethnicity_standardiser):
+        output_data = []
+        for raw_ethnicity in raw_ethnicities:
+            standard_ethnicity = ethnicity_standardiser.standardise(raw_ethnicity)
+            classification_data_item = self.__get_data_item_for_standard_ethnicity(standard_ethnicity)
+            output_data.append(
+                {
+                    "raw_value": raw_ethnicity,
+                    "standard_value": standard_ethnicity,
+                    "display_value": classification_data_item.display_ethnicity,
+                    "parent": classification_data_item.parent,
+                    "order": classification_data_item.order,
+                }
+            )
+        return output_data
+
+    def __get_data_item_for_standard_ethnicity(self, standard_ethnicity):
+        display_ethnicity = self.standard_value_to_display_value_map[standard_ethnicity]
+        return self.classification_data_items[display_ethnicity]
+
+    @staticmethod
+    def get_custom_data_outputs(raw_ethnicities):
+        custom_classification = EthnicityClassification.__get_custom_classification(raw_ethnicities)
+        custom_standardiser = EthnicityClassification.__get_custom_standardiser(raw_ethnicities)
+        return custom_classification.get_outputs(raw_ethnicities, custom_standardiser)
+
+    @staticmethod
+    def __get_custom_classification(raw_ethnicities):
+        classification = EthnicityClassification("custom", "[Custom]")
+
+        unique_raw_values = EthnicityClassification.__order_preserving_remove_duplicates(raw_ethnicities)
+        for ind, value in enumerate(unique_raw_values):
+            classification_data_item = EthnicityClassificationDataItem(
+                display_ethnicity=value, parent=value, order=ind, required=True
+            )
+            classification.add_data_item_to_classification(value, classification_data_item)
+        return classification
+
+    @staticmethod
+    def __get_custom_standardiser(raw_ethnicities):
+        standardiser = EthnicityStandardiser()
+        for ethnicity in raw_ethnicities:
+            standardiser.add_conversion(ethnicity, ethnicity)
+        return standardiser
+
+    @staticmethod
+    def __order_preserving_remove_duplicates(values):
+        unique = []
+        existing = set()
+        for value in values:
+            if value not in existing:
+                existing.add(value)
+                unique.append(value)
+        return unique
```

Comment:
FYI: In (c)Python 3.6+, dicts preserve order by default. So this could be turned into: `return list(dict.fromkeys(values))`.

Reply:
thanks